src/ui/Card.tsx

```diff
@@ -170,15 +174,29 @@ export function Card(props: CardProps): JSX.Element {
   const [clock] = useState(new Clock());
   const panDrawX = useAnimation(panX, velocityX, clock, panState);
 
+  const onCardOffScreen = useCallback(props.onCardOffScreen, []);
+
   useEffect(() => {
     // eslint-disable-next-line @typescript-eslint/no-floating-promises
     (async (): Promise<void> => {
       const messageIds = (await fetchMessageIdsAndLabels(props.threadId))
         .messages;
+
       if (messageIds !== undefined) {
         setMessages(messageIds.map((x) => new Message(x)));
         const messageIdsToFetch = [defined(messageIds[0].id)];
         if (messageIds.length > 1) {
+          let hasMessageInInbox = false;
+          for (const messageId of messageIds) {
+            if (messageId.labelIds?.includes('INBOX')) {
+              hasMessageInInbox = true;
+              break;
+            }
+          }
+          if (!hasMessageInInbox) {
+            await modifyThread(props.threadId, [], ['INBOX']);
+            onCardOffScreen({removeThreadId: props.threadId});
```

Comment:
This seems gross, but I'm not sure what the alternative is, other than grabbing the messageIds earlier in the pipeline.

Reply:
Eventually we'll be fetching threads out of band and actually fetching all their message IDs in an async process ahead of time.

For now, how about if we change fetchThreads to be a generator style thing where it returns one thread at a time but calls fetchMessageIdsAndLabels before returning the thread. So, basically, threads would always be guaranteed to have messageIds and labels. We'd render the first card with 2 round trips in the common case (fetchThreads, fetchMessageIdsAndLabels), and each subsequent card with another fetchMessageIdsAndLabels. 

The UX would be to see 1 card immediately, and then the rest of the cards incrementally show up below that card as we do the fetches until we've eventually fetched all the messageIds and labels.

This would also take fetchMessageIdsAndLabels out of Card, which would be nice. Actually, we could move the two callsed to fetchMessagesById as well, and then Card doesn't need to know anything about gapi and we'd always have the messageIds and the message contents we need for Card. Eventually, we'll store the IDs in Realm and the message contents in local storage, so this works incrementally in that direction I think.

Reply:
That roughly makes sense. I'll give that a shot when I get the chance!